presto-main/src/main/java/com/facebook/presto/operator/PagesRTreeIndex.java

```diff
@@ -0,0 +1,154 @@
+
+public class PagesRTreeIndex
+        implements PagesSpatialIndex
+{
+    private static final long[] EMPTY_ADDRESSES = new long[0];
+
+    private final List<Type> types;
+    private final List<Integer> outputChannels;
+    private final List<List<Block>> channels;
+    private final STRtree rtree;
+    private final BiPredicate<OGCGeometry, OGCGeometry> spatialRelationshipTest;
+    private final JoinFilterFunction filterFunction;
+
+    public static final class GeometryWithAddress
+    {
+        private static final int INSTANCE_SIZE = ClassLayout.parseClass(GeometryWithAddress.class).instanceSize();
+        private final OGCGeometry ogcGeometry;
+        private final long address;
+
+        public GeometryWithAddress(OGCGeometry ogcGeometry, long address)
+        {
+            this.ogcGeometry = ogcGeometry;
+            this.address = address;
+        }
+
+        public OGCGeometry getOgcGeometry()
+        {
+            return ogcGeometry;
+        }
+
+        public long getAddress()
+        {
+            return address;
+        }
+
+        public long getEstimatedMemorySizeInBytes()
+        {
+            return INSTANCE_SIZE + GeometryUtils.getEstimatedMemorySizeInBytes(ogcGeometry);
+        }
+    }
+
+    public PagesRTreeIndex(
+            Session session,
+            LongArrayList addresses,
+            List<Type> types,
+            List<Integer> outputChannels,
+            List<List<Block>> channels,
+            STRtree rtree,
+            BiPredicate<OGCGeometry, OGCGeometry> spatialRelationshipTest,
+            Optional<JoinFilterFunctionFactory> filterFunctionFactory)
+    {
+        this.types = types;
+        this.outputChannels = outputChannels;
+        this.channels = requireNonNull(channels, "channels is null");
+        this.rtree = requireNonNull(rtree, "rtree is null");
+        this.spatialRelationshipTest = requireNonNull(spatialRelationshipTest, "spatial relationship is null");
+        this.filterFunction = filterFunctionFactory.map(factory -> factory.create(session.toConnectorSession(), addresses, channels)).orElse(null);
+    }
+
+    private static Envelope getEnvelope(OGCGeometry ogcGeometry)
+    {
+        com.esri.core.geometry.Envelope env = new com.esri.core.geometry.Envelope();
+        ogcGeometry.getEsriGeometry().queryEnvelope(env);
+
+        return new Envelope(env.getXMin(), env.getXMax(), env.getYMin(), env.getYMax());
+    }
+
+    // Returns an array of addresses from PagesIndex.valueAddresses
+    @Override
+    public long[] findJoinAddresses(int probePosition, Page probe, int probeGeometryChannel)
+    {
+        Block probeGeometryBlock = probe.getBlock(probeGeometryChannel);
+        if (probeGeometryBlock.isNull(probePosition)) {
+            return EMPTY_ADDRESSES;
+        }
+
+        Slice slice = probeGeometryBlock.getSlice(probePosition, 0, probeGeometryBlock.getSliceLength(probePosition));
+        OGCGeometry probeGeometry = deserialize(slice);
```

Comment:
I see that `deserialize` can return null. Can `probeGeometry` be null at all? If no, you may want to verify it.

Reply:
`deserialize` returns `null` only if input is `null`.  Here we know that input is not `null` because of the check above.